crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -156,6 +157,11 @@ pub struct DiskFileSystem {
     invalidator_map: Arc<InvalidatorMap>,
     #[turbo_tasks(debug_ignore, trace_ignore)]
     dir_invalidator_map: Arc<InvalidatorMap>,
+    /// Lock that makes invalidation atomic. It will keep a write lock during
+    /// watcher invalidation and a read lock during other operations.
+    #[turbo_tasks(debug_ignore, trace_ignore)]
+    #[serde(skip)]
+    invalidation_lock: Arc<RwLock<()>>,
```

Comment:
It feels wrong that this should be a lock on `()`. Can we use the type system to ensure we can't do this wrong instead? e.g. by having this lock condition access to the other locks?

Reply:
I think you're suggesting that we have `RwLock<InvalidatorMap>`? Which I was going to suggest, too.